canopen_402/src/motor.cpp

```diff
@@ -481,7 +481,7 @@ void Motor402::handleInit(LayerStatus &status){
         control_word_ = 0;
         start_fault_reset_ = true;
     }
-    if(!switchState(status, State402::Operation_Enable)){
+    if(!switchState(status, no_mode_state_)){
```

Comment:
This should only be done of no mode was selected.

Reply:
It's hard for me to anticipate the purpose of that line. Is Operation_Enabled required for the Homing mode? Or does is support the case where Homing is disabled? To be clear: switchMode does switch to Operation_Enabled anyway.
A propsal:
```
// if(!switchState(status, no_mode_state_)){
//   status.error("Could not enable motor");
//   return;
//}

ModeSharedPtr m = allocMode(MotorBase::Homing);
if(!m){
    if(!switchMode(status, No_Mode)){
        status.error("Could not enable motor");
    }
    return; // homing not supported
}
```
(with no_mode_state handled in switchMode)

Reply:
> with no_mode_state handled in switchMode

:+1: 

But I think no_mode needs to be handled in more places, e.g., https://github.com/ros-industrial/ros_canopen/blob/162c0a8b20fda2f2a4b13ab1c26e1437beeacdb3/canopen_402/src/motor.cpp#L417

Reply:
Is that line required at all? It prevents the operation mode specific bits from being set when state is not Operation_Enabled. But a check if a mode is selected is already in place
https://github.com/ros-industrial/ros_canopen/blob/162c0a8b20fda2f2a4b13ab1c26e1437beeacdb3/canopen_402/src/motor.cpp#L421

In no mode case that won't do anything no matter what state we're in. Or am I missing something?

Reply:
Please have a look at 1a82bcc.
If you don't like what I did to setTransition, the same thing would be possible with the old signature, but I think it's cleaner this way.
If you think handling no_mode_state in handleWrite and/or handleHalt is required, I can add that too.

Reply:
I guess, we need a mechanism like `nextStateForEnabling`  for all possible options for `no_mode_state` (or all).

Reply:
Which state is okay for Nanotec? `Switched_On`?
This might even be a senseful default choice..

Reply:
> Which state is okay for Nanotec?

From [Nanotec PD4-C documentation](https://en.nanotec.com/products/manual/PD4C_CAN_EN/general%252Fds402_power_state_machine.html/):

- Switch on disabled
- Ready to switch on
- Switched on

`switched_on` would be a good default choice! Can you verify if that works with other motors, too?